compiler/src/dotty/tools/backend/sjs/JSCodeGen.scala

```diff
@@ -3111,6 +3124,662 @@ class JSCodeGen()(implicit ctx: Context) {
+         *     scala.runtime.BoxedUnit.UNIT
+         *   }
+         * so that it fits the shape of if/elses we can deal with.
+         */
+        case Block(List(If(cond, thenp, elsep)), s: Select)
+          if s.symbol == defn.BoxedUnit_UNIT =>
+          def block(t: Tree) = Block(t :: Nil, s).withType(s.tpe).withSpan(t.span)
+          js.If(genExpr(cond), genBody(block(thenp)), genBody(block(elsep)))(resultType)
+
+        case _ =>
+          genStatOrExpr(body, isStat)
+      }
+
+      pat match {
+        case lit: Literal =>
+          clauses = (List(genExpr(lit)), genBody(body)) :: clauses
+        case Ident(nme.WILDCARD) =>
+          optElseClause = Some(genBody(body match {
+            case Labeled(_, rhs) if hasSynthCaseSymbol(body) => rhs
+            case _ => body
+          }))
+        case Alternative(alts) =>
+          val genAlts = {
+            alts map {
+              case lit: Literal => genExpr(lit)
+              case _ =>
+                throw new FatalError("Invalid case in alternative in switch-like pattern match: " +
+                  tree + " at: " + tree.sourcePos)
+            }
+          }
+          clauses = (genAlts, genBody(body)) :: clauses
+        case _ =>
+          throw new FatalError("Invalid case statement in switch-like pattern match: " +
+            tree + " at: " + tree.sourcePos)
+      }
+    }
+
+    val elseClause = optElseClause.getOrElse(
+      throw new AssertionError("No elseClause in pattern match"))
+
+    /* Builds a `js.Match`, but simplifies it to a `js.If` if there is only
+     * one case with one alternative, and to a `js.Block` if there is no case
+     * at all. This happens in practice in the standard library. Having no
+     * case is a typical product of `match`es that are full of
+     * `case n if ... =>`, which are used instead of `if` chains for
+     * convenience and/or readability.
+     *
+     * When no optimization applies, and any of the case values is not a
+     * literal int, we emit a series of `if..else` instead of a `js.Match`.
+     * This became necessary in 2.13.2 with strings and nulls.
+     */
+    def buildMatch(cases: List[(List[js.Tree], js.Tree)],
+                   default: js.Tree, tpe: jstpe.Type): js.Tree = {
+
+      def isInt(tree: js.Tree): Boolean = tree.tpe == jstpe.IntType
+
+      cases match {
+        case Nil =>
+          /* Completely remove the Match. Preserve the side-effects of
+           * `genSelector`.
+           */
+          js.Block(exprToStat(genSelector), default)
+
+        case (uniqueAlt :: Nil, caseRhs) :: Nil =>
+          /* Simplify the `match` as an `if`, so that the optimizer has less
+           * work to do, and we emit less code at the end of the day.
+           * Use `Int_==` instead of `===` if possible, since it is a common
+           * case.
+           */
+          val op =
+            if (isInt(genSelector) && isInt(uniqueAlt)) js.BinaryOp.Int_==
+            else js.BinaryOp.===
```

Comment:
This is what would cause a `===` to be emitted instead of the correct `Int_==`. The `===` case should be used only when we are switching on `String`s (including `null` values). In scalac, patmat makes sure that only primitive `Int`s or `java.lang.String`s can be left in `Match` and leave the back-end. That test therefore uses `===` only for strings in scalac. Apparently in dotty other kinds of numeric types survive here, like `Char`s. Either the back-end should convert them to ints, or PatMat should convert them to Ints.

In any case, here we should end up using `===` only for `String`s and `null`. Everything should use `Int_==`.